src/edu/umass/cs/primarybackup/PrimaryBackupManager.java

```diff
@@ -858,19 +881,46 @@ private boolean initializePrimaryEpoch(String groupName, String placementMetadat
         // specified in the placement metadata.
         if (this.myNodeID.toString().equals(preferredCoordinatorNodeId)) {
 
+	    // Try to be coordinator
+	    int count = 0;
+	    while (!this.paxosManager.getPaxosCoordinator(groupName).equals(this.myNodeID)) {
+		if (++count > 10) {
+		    throw new RuntimeException(String.format(
+			"%s:PrimaryBackupManager.initializePrimaryEpoch() - unable to become paxos coordinator for %s:%d after %d tries.",
+			this.myNodeID, groupName, placementEpoch, count
+		    ));
+		};
+
+		this.paxosManager.tryToBePaxosCoordinator(groupName);
+		try {
+		    Thread.sleep(3000);
+		} catch (InterruptedException e) {
+		    throw new RuntimeException(e);
+		}
+	    }
             // FIXME: we need to wait for other replicas (majority) to active first,
             //  before we can propose something.
+	    /*
             try {
                 Thread.sleep(1000);
             } catch (InterruptedException e) {
                 throw new RuntimeException(e);
             }
+	    */
+
+	    // Get IP Addresses of all nodes for rsync (non-deterministic initialization)
+	    Map<String, InetAddress> ipAddresses = new HashMap<>();
+	    NodeConfig<NodeIDType> initNodeConfig = this.messenger.getNodeConfig();
+	    nodes.forEach(node -> ipAddresses.put(
+		String.valueOf(node).toLowerCase(),
+		initNodeConfig.getNodeAddress(node)
+	    ));
 
             // System.out.printf(">> %s Initializing primary epoch for %s\n", myNodeID, groupName);
-            PrimaryEpoch<NodeIDType> zero = new PrimaryEpoch<>(myNodeID, 0);
+            PrimaryEpoch<NodeIDType> epoch = new PrimaryEpoch<>(myNodeID, placementEpoch);
```

Comment:
placementEpoch is not the same as primaryEpoch, primaryEpoch should always be initialized with zero, unless there is another primary election.

- primaryEpoch is increased when we have a new primary in the same replica group.
- placementEpoch is increased when we change the replica group (e.g., {AR1, AR2, AR3} into {AR7, AR8, AR9}).